### PR TITLE
=cls #18287 failed: ClusterShardingSpec ddata mode

### PR DESCRIPTION
**Pull request by oseval**

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

**Comment by patriknw** [comment]
Refs #18287

**Comment by patriknw** [comment]
@smlin Thanks for looking into the issue. It's not a trivial problem and will do some thinking about it tomorrow.

**Comment by patriknw** [comment]
@smlin I looked at the logs from the first failure. After crashing first, the coordinator is started at the fifth node. It gets the initial `GetSuccess` from the replicator and activates itself:

    Sharding Coordinator was moved to the active state State(Map(),Map(),Set(),Set())

State is empty. I agree that that can be because the second node is already crashed and the third and fourth node are not full Up members yet, i.e. the first did not replicate the shard locations in the previous test step to any other node (it was only on first). Then we crash first and the information is lost, i.e. empty state.

I agree that ignoring Register messages from nodes that are not full members yet makes sense.

I might want to add an additional safety by delaying the initial getState until it has received a the initial `CurrentClusterState` and made sure that replica count via `Replicator.ReplicaCount` is >= members.size.
I will do that in a separate pull request. I have a few other things to improve in ClusterSharding anyway.

Let's merge this as the first step.

**Comment by smlin** [comment]
@patriknw I have some thoughts about the coordinator:
1) The Coordinator is just a judge and not a sharding state holder. 
2) The Coordinator could not performs its duties when it does not have full info about all regions  
3) The Coordinator could recover all state from the regions instead of any storage, but it must await to get info from all nodes/regions before moving to the active state. 

I think possible to realize a synchronization of the coordinator with all nodes/regions before moving to the active state. But if one or more nodes temporarily unavailable then such synchronization will be is a long process and during synchronization all cluster-sharding functionality will be locked. 

I don't know how much is bad doing such synchronization on a recovering of the coordinator.

**Comment by patriknw** [comment]
That is a rather different design, with its own set of problems. I don't have the bandwidth to see all implications of that now.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 1]
```diff
@@ -403,23 +394,33 @@ abstract class ShardCoordinator(typeName: String, settings: ClusterShardingSetti
...
-      aliveRegions += region
-      if (state.regions.contains(region))
-        region ! RegisterAck(self)
-      else {
-        gracefulShutdownInProgress -= region
-        update(ShardRegionRegistered(region)) { evt ⇒
-          val firstRegion = state.regions.isEmpty
-          state = state.updated(evt)
-          context.watch(region)
+      if (region.path.address.hostPort == self.path.address.hostPort || members.exists(_.hostPort == region.path.address.hostPort)) {
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: why use `hostPort`?

    if (region.path.address == self.path.address || members.contains(region.path.address)) {

Comment (smlin) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: I was not sure that would be working properly in all cases. Changed.

**Comment by patriknw** [comment]
LGTM

**Comment by patriknw** [comment]
@smlin thanks for contributing, again

**Comment by patriknw** [comment]
Refs #18287

**Pull request merged**